garak/configurable.py

```diff
@@ -0,0 +1,82 @@
+import logging
+import inspect
+from garak import _config
+from garak import _plugins
+
+
+class Configurable:
+    def _load_config(self, config_root=_config):
+        if hasattr(self, "loaded"):
+            return  # only load once, this will ensure the config is not rerun for extending classes
+        local_root = (
+            config_root.plugins if hasattr(config_root, "plugins") else config_root
+        )
+        namespace_parts = self.__module__.split(".")
+        # last part is the namespace, second to last is the plugin type
+        # this will support something like:
+        # plugins['detectors'][x]['generators']['rest']['RestGenerator']
+        # plugins['detectors'][x]['generators']['rest']
+        # plugins['probes'][y]['generators']['rest']['RestGenerator']
+        if len(namespace_parts) > 2:
+            # example class expected garak.generators.huggingface.Pipeline
+            # spec_type = generators
+            # namespace = huggingface
+            # classname = Pipeline
+            # current expected spec_type values are `_plugins.PLUGIN_TYPES`
+            spec_type = namespace_parts[-2]
+            namespace = namespace_parts[-1]
+            classname = self.__class__.__name__
+            plugins_config = {}
+            if isinstance(local_root, dict) and spec_type in local_root:
+                plugins_config = local_root[spec_type]
+            elif hasattr(local_root, spec_type):
+                plugins_config = getattr(local_root, spec_type)
+            if namespace in plugins_config:
+                # example values:
+                # generators: `nim/openai/huggingface`
+                # probes: `dan/gcg/xss/tap/promptinject`
+                attributes = plugins_config[namespace]
+                namespaced_klass = f"{namespace}.{classname}"
+                self._apply_config(attributes)
+                if classname in attributes:
+                    self._apply_config(attributes[classname])
+                elif namespaced_klass in plugins_config:
+                    # for compatibility remove after
+                    logging.warning(
+                        f"Deprecated configuration key found: {namespaced_klass}"
+                    )
+                    self._apply_config(plugins_config[namespaced_klass])
+        self._apply_missing_instance_defaults()
+        self.loaded = True
+
+    def _apply_config(self, config):
+        classname = self.__class__.__name__
+        init_params = inspect.signature(self.__init__).parameters
+        for k, v in config.items():
+            if k in _plugins.PLUGIN_TYPES or k == classname:
```

Comment:
Example:
```
plugins:
  generators:
    huggingface:
      dtype: general
      gpu: 0
      Pipeline:
        dtype: bfloat16
```

When processing `generators["huggingface"]` to build a `huggingface.Pipeline` object we don't want to set `self.Pipeline` to the dictionary value in `generators["huggingface"]["Pipeline"]`.  Since the python naming convention deters capitalization for instance variable names and promotes capitalization for the name of a class reserving the key of the current class made the apply logic reusable for processing `generators["huggingface"]` and subsequently processing ``generators["huggingface"]["Pipeline"]`.